tools/cli/helpers/doc-parser/src/class-doc-parser.php

```diff
@@ -104,15 +146,304 @@ function ( $item ) use ( $ignore ) {
+					);
+
+					$parameters[] = (string) $paramTag . PHP_EOL;
+				}
+
+				foreach (
+					array(
+						'@since',
+						'@module',
+						'@deprecated',
+						'@see',
+						'@uses',
+						'@link',
+						'@type',
+					) as $tagType
+				) {
+					$sinceTags = $phpDocNode->getTagsByName( $tagType );
+					foreach ( $sinceTags as $sinceTag ) {
+						$block['doc']['tags'][] = array(
+							'name'    => substr( $tagType, 1 ),
+							'content' => (string) $sinceTag->value,
+						);
+					}
+				}
+			}
+
+			$filepath = ltrim( substr( $file, strlen( $path ) ), DIRECTORY_SEPARATOR );
+			$blocks[] = array(
+				'path'    => $filepath,
+				'root'    => $path,
+				'classes' => array(
+					array(
+						'methods' => array(
+							array(
+								'hooks' => $file_blocks,
+							),
+						),
+					),
+				),
+			);
+		}
 
 		if ( 'json' === $format ) {
-			$output = json_encode( $output, JSON_PRETTY_PRINT );
+			$blocks = json_encode( $blocks, JSON_PRETTY_PRINT );
+		}
+
+		return $blocks;
+	}
+
+	/**
+	 * Returns all param tag nodes. Tries to recover any parsing errors because of invalid markup.
+	 *
+	 * @param PhpDocNode $phpdocnode the parsed PHPDoc node.
+	 * @return array an array of PhpDocTagNode objects.
+	 */
+	public function get_param_tag_nodes( PhpDocNode $phpdocnode ): array {
+		$tags = $phpdocnode->getParamTagValues();
+
+		// Looking for invalid param tags.
+		foreach ( $phpdocnode->getTags() as $tag ) {
+			'@phan-var \PHPStan\PhpDocParser\Ast\PhpDoc\PhpDocTagNode $tag';
+
+			if ( $tag->name === '@param' && $tag->value instanceof InvalidTagValueNode ) {
+				$tag_value = $tag->value;
+				'@phan-var InvalidTagvalueNode $tag_value';
+
+				$pieces      = explode( ' ', $tag_value->value );
+				$type        = new IdentifierTypeNode( $pieces[0] );
+				$name        = $pieces[1] ?? 'argument';
+				$description = implode( ' ', array_slice( $pieces, 2 ) );
+				$tags[]      = new ParamTagValueNode( $type, true, $name, $description, false );
+			}
+		}
+
+		return $tags;
+	}
+
+	/**
+	 * Returns a list of PHP files in a folder, recursing into subfolders. Heavily inspired by
+	 * the WordPress PHPDoc parser.
+	 *
+	 * @see https://github.com/WordPress/phpdoc-parser/blob/7fc2227d2d4fb73f9f0b6e233413f3f9f9840e80/lib/runner.php#L17
+	 *
+	 * @param string $directory the folder to look in.
+	 *
+	 * @return array an array of filenames.
+	 * @throws \Exception $e If unable to traverse the filesystem.
+	 */
+	public function get_wp_files( $directory ) {
+		$iterableFiles = new \RecursiveIteratorIterator(
+			new \RecursiveDirectoryIterator( $directory )
+		);
+		$files         = array();
+
+		try {
+			foreach ( $iterableFiles as $file ) {
+				if ( 'php' !== $file->getExtension() ) {
+					continue;
+				}
+
+				$files[] = $file->getPathname();
+			}
+		} catch ( \UnexpectedValueException $exc ) {
+			throw new \Exception( sprintf( 'Directory [%s] contained a directory we can not recurse into', $directory ) );
```

Comment:
This exception message is kind of useless. It might be better to just let the original exception be uncaught. Or at the least do something like
```suggestion
			throw new \Exception( sprintf( 'Directory [%s] contained a directory we can not recurse into', $directory ), 0, $exc );
```
so we can see the original exception too.

Also personally I'd probably use a typed exception (like UnexpectedValueException) instead of throwing a plain Exception. But in the end it doesn't make much difference here.